### ISSUE TITLE
Params incorrect for form submitted for nested resource 

### ISSUE THREAD

**Issue by helphop** [hopsoft/stimulus_reflex#290]
# Bug Report

## Bug Description 

When I submit the form using without StimulusReflex all of the parameter data is present including:

```
id: 6
post_id: "photos-reveal-devastation-from-oil-spill-in-mauritius"
```

```

<ActionController::Parameters {"_method"=>"patch", "authenticity_token"=>"MDtxGtFe65lUnSZ8K32PdwNNBv7aAlOAmkAhjkKio7IrMTOdfMmn1s6M6vTsMarWRiBVmoUips7El17UxgToYw==", "element"=>{"content"=>"<h2><strong>Enim nulla aliquet porttitor lacus luctus accumsan.&nbsp;</strong></h2><div><br>Cursus euismod quis viverra nibh cras pulvinar mattis nunc. Urna id volutpat lacus laoreet non. Ut tortor pretium viverra suspendisse. something</div>"}, "commit"=>"Save", "controller"=>"authors/elements", "action"=>"update", "post_id"=>"photos-reveal-devastation-from-oil-spill-in-mauritius", "id"=>"6"} permitted: false>

```

However, when I submit a form using StimulusReflex the params that are submitted along with the form are including different data and are missing the post_id: 

```
id: "photos-reveal-devastation-from-oil-spill-in-mauritius"
no post_id present
```

```

<ActionController::Parameters {"controller"=>"authors/posts", "action"=>"edit", "id"=>"photos-reveal-devastation-from-oil-spill-in-mauritius", "_method"=>"patch", "authenticity_token"=>"oBnP815mQ/E0IltrBnc4BB5CQurRxiA4o9qZTVE3dWj0tLy+krC2nN+Sy8+sAvN9gueYsx87YBc2YxQTrScJdg==", "element"=>{"content"=>"<h2><strong>Enim nulla aliquet porttitor lacus luctus accumsan.&nbsp;</strong></h2><div><br>Cursus euismod quis viverra nibh cras pulvinar mattis nunc. Urna id volutpat lacus laoreet non. Ut tortor pretium viverra suspendisse. something</div>"}} permitted: false>

```

I have the following in my routes.rb file:
```
  scope module: 'authors' do
    resources :posts do
      resources :elements
    end
  end
```

I have a Post made up of many elements:
```
class Post < ApplicationRecord
  has_many :elements
end

class Element < ApplicationRecord
 belongs_to :post 
end

```

On my edit post page I have one form for each of the elements that comprise the post:
views/posts/edit.html.erb

```

<%= @post.element.each do |element| %>
  <%= render partial: 'paragraph_form', locals: { post: @post, element: element} %>
<% end %>

```

Paragraph Form

```
<%= form_with(model: [post,element], data: { reflex: "submit->ElementsReflex#update", signed_id: element.to_sgid.to_s})  do |form| %>
  <%= form.rich_text_area :content %>
  <div class="mt-3 d-flex justify-content-between">
    <div>
      <%= form.submit "Save", class: 'btn btn-primary' %>
      <a href="#" class="cancel btn btn-secondary">Cancel</a>
    </div>
      <%= link_to 'Delete', post_element_path(element.post, element), method: :delete, class: 'btn btn-danger', data: { confirm: 'Are you sure?'} %>
  </div>
<% end %>
```

This resolves correctly and I have the following in my form (simplified):

```
<form data-reflex="submit->ElementsReflex#update" action="/posts/new-titles/elements/16"  data-remote="true" method="post" data-action="submit->elements#__perform">
```

Here is the full output from the submission:

`
StimulusReflex::Channel#receive(
{"target"=>"ElementsReflex#update", "args"=>[], "url"=>"http://localhost:3000/posts/new-titles/edit", "attrs"=>{
"data-reflex"=>"submit->ElementsReflex#update", 
"data-signed-id"=>"BAh7CEkiCGdpZAY6BkVUSSIcZ2lkOi8vcy1ibG9nL0VsZW1lbnQvMTYGOwBUSSIMcHVycG9zZQY7AFRJIgxkZWZhdWx0BjsAVEkiD2V4cGlyZXNfYXQGOwBUSSIdMjAyMC0wOS0xM1QxNzoyNDo0Ny40NjdaBjsAVA==--707039c8db2558e68dd5af108e93ebdfd2d2c9b1", "action"=>"/posts/new-titles/elements/16", "accept-charset"=>"UTF-8", "data-remote"=>"true", "method"=>"post", "data-action"=>"submit->elements#__perform", "checked"=>false, "selected"=>false, "tag_name"=>"FORM"}, "dataset"=>{"data-reflex"=>"submit->ElementsReflex#update", "data-signed-id"=>"BAh7CEkiCGdpZAY6BkVUSSIcZ2lkOi8vcy1ibG9nL0VsZW1lbnQvMTYGOwBUSSIMcHVycG9zZQY7AFRJIgxkZWZhdWx0BjsAVEkiD2V4cGlyZXNfYXQGOwBUSSIdMjAyMC0wOS0xM1QxNzoyNDo0Ny40NjdaBjsAVA==--707039c8db2558e68dd5af108e93ebdfd2d2c9b1", "data-remote"=>"true", "data-action"=>"submit->elements#__perform"}, "selectors"=>[], "reflexId"=>"e687bf2e-e108-4200-adbe-aab85ab99ba5", "permanent_attribute_name"=>"data-reflex-permanent", "params"=>{"_method"=>"patch", "authenticity_token"=>"dSWS/Ezp/wjaVGs4fyPT+rxJBstorYswpWCqk6pdp1HXf2ot+eK5V0/VCmaaKhcwNu7gac8n3dAqno6QYVLptQ==", "element"=>{"content"=>"<div>another one bites the</div>"}}})
`

## To Reproduce

Create a page that contains several forms. The forms are for the has_many model of the association.
Submit the form to a Stimulus Reflex reflex and examine the params object.

## Expected behavior

I would expect that the params submitted via a Reflex would be the same as if I submitted the form to a Rails controller.

### StimulusReflex

- Gem: stimulus_reflex 3.2.3
- Node package:  "stimulus_reflex": "^3.3.0-pre2",

### External tools

- Ruby:ruby 2.6.5p114
- Rails: 6.0.3.2
- Node: v14.5.0

### Browser

- Browser Firefox  
- Version 79

**Comment by leastbad** [comment]
Hi Mitchell, we're looking into this. Are you on Discord, by any chance?

**Comment by RolandStuder** [comment]
There is a bug with the params, that if you use params in your reflex you will be missing path and query parameters. I think in your case the `post_id` is a path parameter, so this bug might be what the problem is.

So @helphop could you try by updating the gem to 3.3.0-pre2 

The bug is fixed but only in the gem in the prerelease versions. 
Here is the fix: https://github.com/hopsoft/stimulus_reflex/pull/249.

Could you check if this fixes your issue?

**Comment by helphop** [comment]
Could someone have a look at this. I've also found that when I send a file_field to a reflex the file_field parameter and data are not sent.

**Comment by helphop** [comment]
Thanks very much for those details and especially for  the link to documentation for file uploads. I made the error of searching the documents for ActiveStorage instead of file uploads which is why I did not find it.  My mistake.  I will try out the form-serialize suggestion and see what the data structure returns.

**Comment by helphop** [comment]
@leastbad I've tried as instructed to investigate the via importing the form-serialize js into a Stimulus controller and examining the data from form-serialize.  It's too complicated for my expertise to be able to make any sense of all of the information.  I'm just not skilled enough to respond with anything helpful. I'll just submit my forms using the standard Rails methodology until I understand better. 

Even when I try to use a callback beforeReflex it won't fire.  Seems this is just too complex for my abilities.

**Comment by leastbad** [comment]
I'm sorry to hear that, Mitchell, but I understand. Don't be so hard on yourself, either way. I appreciate you checking it out. All of us are learning in real-time.

I'd offer to help (for example, you mentioned beforeReflex not firing) but I can't debug an anecdote. If you are able to put your project on Github - it doesn't have to be public, you can add me as a collaborator - I'm happy to take a look and see if I can help get you back on track.

**Comment by helphop** [comment]
Hey thanks for your support I very much appreciate it.  I have followed your guidance and have created and invited you to this trial app.  

I have also created a reflex that accepts an argument as follows:

```
class ParagraphReflex < ApplicationReflex
before_reflex :set_objects, except: [:update]

  def update(my_params)
    byebug
    if @paragraph.update(paragraph_params)
      render_paragraph
    end
  end
...

```

I have experimenting with submitting a form using a StimulusController  then using the form-serialize to serialize the form data and passing that in with the call to StimulusReflex as follows:

```
import { Controller } from 'stimulus'
import StimulusReflex from 'stimulus_reflex'
export default class extends Controller {

  connect() {
    StimulusReflex.register(this)
  }

  update(element) {
    var serialize = require("form-serialize");
    let my_params = serialize(element.target, { hash: true, empty: true})
    this.stimulate("Paragraph#update", element, my_params)
  }
}
```

The params and my_params are slightly different:

params:

`<ActionController::Parameters {"controller"=>"pages", "action"=>"show", "notebook_id"=>"3", "id"=>"1", "_method"=>"patch", "authenticity_token"=>"1JeJgqusIAEVnldgaLdW/dsdXu0keu+I8tED/thuSjNmweEYAI7N1KowDlB1wFG+QI+yJM/toZDHNQL5OK/eGg==", "paragraph"=>{"content"=>"<div>A new message again</div>"}} permitted: false>`

my_params

`{"_method"=>"patch", "authenticity_token"=>"1JeJgqusIAEVnldgaLdW/dsdXu0keu+I8tED/thuSjNmweEYAI7N1KowDlB1wFG+QI+yJM/toZDHNQL5OK/eGg==", "paragraph"=>{"content"=>"<div>A new message again</div>"}}`

In both cases they are missing form data.  In params they are missing the paragraph id.  
my_params is missing all of the id's.

So it seems that submitting the form using this JS library is not 100% compatible with Rails as you noted earlier.

**Comment by leastbad** [comment]
I have the app running, but I'm getting an `InvalidAuthenticityToken` exception when I attempt to create a user to login. I even created a credentials file for my development environment with a secret key base.

Any ideas?

**Comment by helphop** [comment]
Possibly put   skip_before_action :verify_authenticity_token inside of Application Controller just to get things up an running. I've had that issue before but don't remember the solution. 
Nathan Jones <notifications@github.com> wrote:
“I have the app running, but I'm getting an InvalidAuthenticityToken exception when I attempt to create a user to login. I even created a credentials file for my development environment with a secret key base.


Any ideas?”

**Comment by leastbad** [comment]
Alright, I'm in. Let me play with this and I'll get back to you.

**Comment by helphop** [comment]
:) Much appreciated
Nathan Jones <notifications@github.com> wrote:
“Alright, I'm in. Let me play with this and I'll get back to you.”

**Comment by RolandStuder** [comment]
@helphop I think this is working/broken as intended.

TLDR: 

The `params` object in a reflex is extremely similar to the params object in a controller action from a form submit. **But** the tricky difference is that **in a reflex the path and query params come from the originating url**, and **in a controller action** we get the **path and query params from the url the form was submitted to**.

So in you reflex you find the right paragraph not by using the params but by passing around the id via the dataset.

You already prepared that in your code (`signed_id`)
`<%= form_with(model: [notebook, page, paragraph], data: {controller: "paragraph", action: "submit->paragraph#update", signed_id: paragraph.to_sgid.to_s}) do |form| %>`

Access it in the reflex with:

`element.dataset[:signed_id]`

---

## Why are you seeing what you are seeing in the params

One needs to understand what the source of the params in a reflex is. And there are two sources:

1. the closest form
2. the url of the currently displayed page

## my_params is the complete serialization of the closest form

```
{"_method"=>"patch", "authenticity_token"=>"1JeJgqusIAEVnldgaLdW/dsdXu0keu+I8tED/thuSjNmweEYAI7N1KowDlB1wFG+QI+yJM/toZDHNQL5OK/eGg==", "paragraph"=>{"content"=>"<div>A new message again</div>"}}
```
In your example my_params is the serialization of the form. The form only contains one field, only (the  content field), along with the method and the authenticity_token.

You say you are missing all the ids. Now usually you get the ids from the path that the form submit to, this information you pass into the form by using `form_with(model: [notebook, page, paragraph]. **But that path has no bearing on the params in the reflex**. It is not used at all by the reflex 

## Why does the reflex have some of the ids then?

```<ActionController::Parameters {"controller"=>"pages", "action"=>"show", "notebook_id"=>"3", "id"=>"1", "_method"=>"patch", "authenticity_token"=>"1JeJgqusIAEVnldgaLdW/dsdXu0keu+I8tED/thuSjNmweEYAI7N1KowDlB1wFG+QI+yJM/toZDHNQL5OK/eGg==", "paragraph"=>{"content"=>"<div>A new message again</div>"}} permitted: false>```

As you note in the reflex you have some more parameter, namely `notebook_id` and `id`. This is the confusing part. These parameters come from the page that rendered the form, it does **not** come from the action-path of the form.

This is just how stimulus reflex works, when you trigger a reflex, the url of the current page gets sent along. In the reflex SR infers the path and query params from the url.

---

## Could we do better?

Basically the situation is  that the params in a reflex are not the same ones you would receive, if you would just submit the form.
I think it might be worth to discuss if some renaming might be in order here. Such as deprecating params and instead making the form data and the path params available with separate methods. I think this would be premature. But I would not be surprised if more people run into this confusion.

**Comment by leastbad** [comment]
Thank you Roland, this is incredibly lucid.

I agree that deprecating `params` is likely a non-starter, but certainly we could improve the documentation around this so people aren't left wondering WTF is going on.

At the end of the day, SR was not designed or developed with forms in mind - Jason realizing that a form could be serialized was closer to a happy accident. It's reasonable for folks to assume that params will behave exactly as a traditional form submission might work, confusing the implementation with the state data it represents.

I suspect that the easiest way to explain it might be to take the approach of showing ways SR `params` are like and not like ActionController `params`. Certainly it couldn't hurt to spell out that the form submit path is not part of the equation, as it seems like much of the potential for confusion starts here.

**Comment by leastbad** [comment]
I'm going to close this for now. Thanks to this thread, there's more thorough documentation on forms.

https://docs.stimulusreflex.com/working-with-forms#form-submission
